Contextual compression retriever

dev2049 opened the pull request.

dev2049 committed:
rfc

dev2049 committed:
Merge branch 'master' into dev2049/contextual_compression

dev2049 committed:
cr

dev2049 commented:
will add tests and move prompt to own file but wanted to get ppl's high level thoughts first

dev2049 committed:
nit

hwchase17 committed:
cr

eyurtsev commented on langchain/retrievers/contextual_compression.py:
```diff
@@ -0,0 +1,111 @@
+        """Filter down documents."""
+
+    @abstractmethod
+    def afilter(self, docs: List[Document], query: str) -> List[Document]:
```
Async?

eyurtsev commented on langchain/retrievers/contextual_compression.py:
```diff
@@ -0,0 +1,111 @@
+
+class BaseDocumentFilter(BaseModel, ABC):
+    @abstractmethod
+    def filter(self, docs: List[Document], query: str) -> List[Document]:
```
Could we use sequence instead of list on arguments

dev2049 commented on langchain/retrievers/contextual_compression.py:
```diff
@@ -0,0 +1,111 @@
+    return {"question": query, "context": doc.page_content}
+
+
+class BaseDocumentFilter(BaseModel, ABC):
```
to me filter sounds like it's just choosing which docs to keep. maybe BaseDocumentCompressor/Compression?

vowelparrot commented on langchain/retrievers/contextual_compression.py:
```diff
@@ -0,0 +1,111 @@
+
+
+def _get_default_chain_prompt() -> PromptTemplate:
+    template = """Given the following question and context, extract any part of the context *as is* that is relevant to answer the question.
```
I'm not convinced it's important,  but a lot of queries aren't "questions" per-se

vowelparrot reviewed (commented):
I like the idea. I'd imagine there may be some use of having a "None" option as well

dev2049 committed:
Merge branch 'dev2049/contextual_compression' of github.com:hwchase17/langchain into dev2049/contextual_compression

dev2049 committed:
add filters and pipeline

dev2049 commented:
A lot of things going on here, added all this to one branch to try to flesh out my own thinking but would split this pr up if/when we agree on which abstractions are actually worth keeping

Added some more DocumentFilters and notion of a filter pipeline. Think these are now really more like document "transformers" than "filters", but can hash out naming later.

There's now filters for 
- llm-based compression (one doc at a time)
- llm-based filtering based on relevance to query
- embedding based filtering based on relevance to query
- embedding based filtering based on redundancy with other docs
- text splitter "filtering" (this is just a wrapper to fit splitters into the pipeline framework)


A  filter pipeline would look something like this
```python
pipeline_filter = DocumentFilterPipeline(
    filters=[
        SplitterDocumentFilter(splitter=CharacterTextSplitter(chunk_size=100, chunk_overlap=0, separator=". ")),
        EmbeddingRedundantDocumentFilter(embeddings=OpenAIEmbeddings()),
        EmbeddingRelevancyDocumentFilter(embeddings=OpenAIEmbeddings(), similarity_threshold=0.8),
    ]
)
```
this pipeline would break down docs into smaller chunks, remove the redundant ones, then keep only the relevant ones. and it'd reuse embeddings between steps 2 and 3

any and all feedback appreciated. very possible this is over-engineered, won't be the least bit offended if you say as much :)

dev2049 committed:
Merge branch 'master' into dev2049/contextual_compression

dev2049 committed:
cr

hwchase17 commented on langchain/retrievers/contextual_compression.py:
```diff
@@ -0,0 +1,54 @@
+            Sequence of relevant documents
+        """
+        docs = self.base_retriever.get_relevant_documents(query)
+        retrieved_docs = [RetrievedDocument.from_document(doc) for doc in docs]
```
hm if we just use retrieved document here only seems a bit weird

it would be nice if normal retrievers could also return RetrievedDocuments

could check if `doc` already retrieved doc, convert if not. and just return retrieved doc?

retrieved doc may be a concept worth living in general schema?

hwchase17 replied:
actually.... if we arent 100% sure about retrieved doc as an abstaction, could be kinda nice to have internal only? in whcih case maybe make it `_RetrievedDocument`?

dev2049 replied:
personally not super convinced it's a long term abstraction, so happy to make it `_RetrievedDocument`

dev2049 committed:
cr

dev2049 committed:
cr

dev2049 commented on langchain/math_utils.py:
```diff
@@ -0,0 +1,20 @@
+Matrix = Union[List[List[float]], List[np.ndarray], np.ndarray]
+
+
+def cosine_similarity(x: Matrix, y: Matrix) -> np.ndarray:
```
if we're ok making sklearn a dependency could just import

dev2049 replied:
otherwise i can add some unit tests

dev2049 committed:
docstring

dev2049 committed:
private

dev2049 committed:
Merge branch 'master' into dev2049/contextual_compression

dev2049 committed:
cosine unit test

dev2049 committed:
unit test

dev2049 committed:
add example nb

dev2049 commented on langchain/vectorstores/utils.py:
```diff
@@ -4,10 +4,7 @@
+from langchain.math_utils import cosine_similarity
 
 
 def maximal_marginal_relevance(
```
should probably add some unit tests for this

dev2049 committed:
fmt

hinthornw reviewed (commented):
.

dev2049 committed:
Merge branch 'master' into dev2049/contextual_compression

dev2049 committed:
rename

dev2049 committed:
cr

dev2049 committed:
Merge branch 'master' into dev2049/contextual_compression

dev2049 committed:
cr

dev2049 committed:
add doc transformer

dev2049 committed:
rename

dev2049 committed:
cr

dev2049 committed:
fix

dev2049 committed:
nit

dev2049 committed:
nb

dev2049 committed:
nb

dev2049 commented on langchain/retrievers/contextual_compression.py:
```diff
@@ -0,0 +1,51 @@
+        """
+        docs = await self.base_retriever.aget_relevant_documents(query)
+        compressed_docs = await self.base_compressor.acompress_documents(docs, query)
+        return list(compressed_docs)
```
is this kosher in async? @agola11 @hwchase17

hwchase17 replied:
sure

hwchase17 committed:
Merge branch 'master' into dev2049/contextual_compression

hwchase17 committed:
Merge branch 'dev2049/contextual_compression' of github.com:hwchase17/langchain into dev2049/contextual_compression

hwchase17 committed:
contextual compression (#3253)

remove useless classes
